src/output-flex.cpp

```diff
@@ -1055,15 +1039,104 @@ void output_flex_t::pending_way(osmid_t id)
     auto &way = m_buffer.get<osmium::Way>(0);
 
     m_context_way = &way;
-    call_lua_function(m_process_way, way);
+    get_mutex_and_call_lua_function(m_process_way, way);
     m_context_way = nullptr;
     m_num_way_nodes = std::numeric_limits<std::size_t>::max();
     m_buffer.clear();
 }
 
+void output_flex_t::mark_relation_members(osmium::Relation const &relation)
+{
+    if (!m_mark_relation_members) {
+        return;
+    }
+
+    std::lock_guard<std::mutex> guard{lua_mutex};
+
+    m_context_relation = &relation;
+    call_lua_function(m_mark_relation_members, relation);
+    m_context_relation = nullptr;
+
+    // If the function returned nil there is nothing to be marked.
+    if (lua_type(lua_state(), -1) == LUA_TNIL) {
+        lua_pop(lua_state(), 1); // return value (nil)
+        return;
+    }
+
+    if (lua_type(lua_state(), -1) != LUA_TTABLE) {
+        throw std::runtime_error{"mark_relation_members() returned something "
+                                 "other than nil or a table"};
+    }
+
+    // We have established that we have a table. Get the 'ways' field...
+    lua_getfield(lua_state(), -1, "ways");
+    int const ltype = lua_type(lua_state(), -1);
+
+    // No 'ways' field, that is okay, nothing to be marked.
+    if (ltype == LUA_TNIL) {
+        lua_pop(lua_state(), 2); // return value (a table), ways field (nil)
+        return;
+    }
+
+    if (ltype != LUA_TTABLE) {
+        throw std::runtime_error{
+            "Table returned from mark_relation_members() contains 'ways' "
+            "field, but it isn't an array table"};
+    }
+
+    // Iterate over the 'ways' table to get all ids...
+    lua_pushnil(lua_state());
+    while (lua_next(lua_state(), -2) != 0) {
+        if (!lua_isnumber(lua_state(), -2)) {
+            throw std::runtime_error{
+                "Table returned from mark_relation_members() contains 'ways' "
+                "field, but it isn't an array table"};
```

Comment:
...array table of numbers

Reply:
The check is for the table key not the value. The table key is a number in "array tables", so we are checking the table is an array, not a hash.